components/catalogs/back/src/Application/Handler/GetMappedProductsHandler.php

```diff
@@ -79,10 +84,38 @@ public function __invoke(GetMappedProductsQuery $query): array
 
         $productMapping = $catalog->getProductMapping();
 
+        $this->hydrateCachedCategoryCodesAndLocales($productMapping, \array_column($products, 'uuid'));
+
         return \array_map(
             /** @param RawProduct $product */
             fn (array $product): array => $this->productMapper->getMappedProduct($product, $productMappingSchema, $productMapping),
             $products,
         );
     }
+
+    /**
+     * @param ProductMapping $productMapping
+     * @param UuidInterface[] $productUuids
+     */
+    private function hydrateCachedCategoryCodesAndLocales(array $productMapping, array $productUuids): void
+    {
+        /** @var string[] */
+        $categoryLocales = \array_reduce(
+            $productMapping,
+            fn (array $carry, array $item): array => $item['source'] === 'categories' ?
+                \array_unique(
+                    \array_merge($carry, [$item['locale']]),
+                )
+                : $carry,
+            [],
+        );
```

Comment:
```suggestion
        $categoryLocales = [];
    
        foreach ($productMapping as $targetSourceAssociation) {
            if ($targetSourceAssociation['source'] === 'categories') {
                $locale = $targetSourceAssociation['locale'];
    
                if (!in_array($locale, $categoryLocales)) {
                    $categoryLocales[] = $locale;
                }
            }
        }
```
Is it much clearer ?

Reply:
@guillaumefer-akeneo not for me. `array_reduce` is a very good tool IMO. But you revealed that In this case, I made several errors. I keep bad names `$carry` and especially`$item`, it's an error. Furthermore, I think I can use your `in_array` test in place of `array_unique`. WDYT about something like this :
```suggestion
        $categoryLocales = \array_reduce(
            $productMapping,
            fn (array $localesList, array $mappingEntry): array => 
                $mappingEntry['source'] === 'categories' && !\in_array($mappingEntry['locale'] , $localesList) 
                ? \array_merge($localesList, [$mappingEntry['locale')])
                : $localesList,
            [],
        );
```

@akeneo/squad-octopus WDYT about `array_reduce` ?

Reply:
Not a big fan to use `array_reduce` in this case because it increases the cognitive complexity (nested functions). In my case, I had to open the doc to understand what the code was doing.

Reply:
I agree with Guillaume, `array_reduce` often results in difficult-to-read code and a `foreach` statement would do a better job IMO
especially when the closure is not a oneliner